core/amber/src/main/scala/edu/uci/ics/amber/engine/architecture/messaginglayer/BatchToTupleConverter.scala

```diff
@@ -39,9 +40,20 @@ class BatchToTupleConverter(
     new mutable.HashMap[LinkIdentity, mutable.HashSet[ActorVirtualIdentity]]
   private val completedLinkIds = new mutable.HashSet[LinkIdentity]()
 
+  // This buffer handles data comes before the UpdateInputLinking control message.
+  // During normal processing, this should not happen. But it may happen during worker recovery.
+  // TODO: remove this buffer once we move the upstream map to DP.
+  private val dataBufferForUnRegisteredInput =
```

Comment:
lets move senderchange, endmarker, and endofall to DP thread, and let main actor directly enqueue raw messages, so recovery is easier.